tls/s2n_client_key_exchange.c

```diff
@@ -31,13 +31,29 @@
 #include "utils/s2n_random.h"
 #include "utils/s2n_safety.h"
 
-#define get_client_hello_protocol_version(conn) (conn->client_hello_version == S2N_SSLv2 ? conn->client_protocol_version : conn->client_hello_version)
-
 typedef S2N_RESULT s2n_kex_client_key_method(const struct s2n_kex *kex, struct s2n_connection *conn, struct s2n_blob *shared_key);
 typedef void *s2n_stuffer_action(struct s2n_stuffer *stuffer, uint32_t data_len);
 
 static int s2n_rsa_client_key_recv_complete(struct s2n_connection *conn, bool rsa_failed, struct s2n_blob *shared_key);
 
+/*
+ *= https://www.rfc-editor.org/rfc/rfc5246#section-7.4.7.1
+ *#   client_version
+ *#      The latest (newest) version supported by the client.  This is
+ *#      used to detect version rollback attacks.
+ *
+ * However, TLS1.2 rsa kex does not account for the existence of TLS1.3.
+ * Therefore "latest" actually means "latest up to TLS1.2".
+ */
+static S2N_RESULT s2n_client_key_exchange_get_rsa_client_version(struct s2n_connection *conn,
+        uint8_t client_version[S2N_TLS_PROTOCOL_VERSION_LEN])
+{
+    uint8_t client_version_for_rsa = MIN(conn->client_protocol_version, S2N_TLS12);
```

Comment:
Why would you use client_protocol_version here? It seems like what you actually want is client_hello.legacy?

Reply:
In practice, either would probably work. But the RFC phrased it as "the latest version supported by the client", which maps in our logic / definitions to "client_protocol_version". So I think using "client_protocol_version" makes more sense.